pandas/core/ops/array_ops.py

```diff
@@ -321,6 +325,16 @@ def comparison_op(left: ArrayLike, right: Any, op) -> ArrayLike:
                 "Lengths must match to compare", lvalues.shape, rvalues.shape
             )
 
+    if (is_string_dtype(lvalues) and is_object_dtype(rvalues)) or (
+        is_object_dtype(lvalues) and is_string_dtype(rvalues)
+    ):
+        if lvalues.dtype.name == "string" and rvalues.dtype == object:
+            lvalues = lvalues.astype("string")
+            rvalues = pd_array(rvalues, dtype="string")
```

Comment:
We might need to do the casting the other way around. Instead of casting the object to string and then compare both as strings, I think we have to cast the string to object and compare both as object dtype.

The reason for this is that casting to string might actually convert values to strings, and then we are no longer doing the comparison for the original values.

```python
>>> ser_string = pd.Series(["1", "b"])
>>> ser_mixed = pd.Series([1, "b"])
>>> ser_string == ser_mixed
0    False
1     True
dtype: bool

>>> ser_string == ser_mixed.astype("string")
0    True
1    True
dtype: bool
```

So if we would do that casting under the hood, the result would change in this case.

Reply:
And we should add this case to the tests!